### PR TITLE
Feature/429 notifications crud backend

### LINKED ISSUE #429: On job finish notifications - CRUD Backend
Explore the following architecture in a PoC
- For every job that finishes, send a email notification to user 
- User should be able to set up notifications on project level

### PR DESCRIPTION
**Pull request by kevinwallimann**
Closes #429

### CODE REVIEW

File: src/main/resources/db_scripts/liquibase/v0.4.3.add-notification-rule.sql [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+create table "notification_rule" (
+  "id" BIGSERIAL NOT NULL PRIMARY KEY,
+  "project" VARCHAR,
+  "workflow_prefix" VARCHAR,
+  "min_elapsed_secs_last_success" BIGINT,
```
Comment (jozefbakus) [inline src/main/resources/db_scripts/liquibase/v0.4.3.add-notification-rule.sql:20]: Question: What if I set it to 0 ? Will it send notification every time ?

File: src/main/scala/za/co/absa/hyperdrive/trigger/models/tables/JdbcTypeMapper.scala [hunk 2]
```diff
@@ -122,4 +119,22 @@ trait JdbcTypeMapper {
     },
     column => column.as[JobInstanceParameters]
   )
+
+  implicit lazy val notificationRuleMapper: JdbcType[NotificationRule] = MappedColumnType.base[NotificationRule, JsValue](
+    notificationRule => Json.toJson(notificationRule),
+    column => column.as[NotificationRule]
+  )
+
+  type Recipients = Seq[String]
```
Comment (jozefbakus) [inline src/main/scala/za/co/absa/hyperdrive/trigger/models/tables/JdbcTypeMapper.scala]: I would move it into NotificationRule class

File: src/main/scala/za/co/absa/hyperdrive/trigger/models/tables/NotificationRuleHistoryTable.scala [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+  import api._
+
+  final class NotificationRuleHistoryTable(tag: Tag) extends Table[NotificationRuleHistory](tag, _tableName = "notification_rule_history") {
+    def id: Rep[Long] = column[Long]("id", O.PrimaryKey, O.AutoInc, O.SqlType("BIGSERIAL"))
+    def changedOn: Rep[LocalDateTime] = column[LocalDateTime]("changed_on")
+    def changedBy: Rep[String] = column[String]("changed_by")
+    def operation: Rep[DBOperation] = column[DBOperation]("operation")
+    def notificationRuleId: Rep[Long] = column[Long]("notification_rule_id")
+    def notificationRule: Rep[NotificationRule] = column[NotificationRule]("notification_rule")
```
Comment (jozefbakus) [inline src/main/scala/za/co/absa/hyperdrive/trigger/models/tables/NotificationRuleHistoryTable.scala]: O.SqlType("JSONB") ?

File: src/main/scala/za/co/absa/hyperdrive/trigger/persistance/NotificationRuleRepository.scala [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+@stereotype.Repository
+class NotificationRuleRepositoryImpl(override val notificationRuleHistoryRepository: NotificationRuleHistoryRepository) extends NotificationRuleRepository {
+
+  import api._
+
+  private val repositoryLogger = LoggerFactory.getLogger(this.getClass)
+
+  private implicit class DBIOActionOps[T](val action: api.DBIO[T]) {
+    def withErrorHandling(errorMessage: String)(implicit ec: ExecutionContext): DBIOAction[T, NoStream, Effect.All] = {
```
Comment (jozefbakus) [inline src/main/scala/za/co/absa/hyperdrive/trigger/persistance/NotificationRuleRepository.scala:52]: Cool

**Review by jozefbakus — COMMENTED** [review]
Looks good but you could have exposed REST endpoints. 
Would be easier to try it.

File: src/main/resources/db_scripts/liquibase/v0.4.3.add-notification-rule.sql (hunk 1 above)
Comment (kevinwallimann) [reply src/main/resources/db_scripts/liquibase/v0.4.3.add-notification-rule.sql:20]: Yes that's the idea

File: src/main/scala/za/co/absa/hyperdrive/trigger/models/tables/NotificationRuleHistoryTable.scala (hunk 3 above)
Comment (kevinwallimann) [reply src/main/scala/za/co/absa/hyperdrive/trigger/models/tables/NotificationRuleHistoryTable.scala]: Hmm, I'm wondering why this worked anyway

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/controllers/NotificationRuleController.scala [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+import scala.compat.java8.FutureConverters._
+import scala.concurrent.ExecutionContext
+import scala.concurrent.ExecutionContext.Implicits.global
+
+@RestController
+class NotificationRuleController @Inject()(notificationRuleService: NotificationRuleService) {
+  private val logger = LoggerFactory.getLogger(this.getClass)
+
+  @PutMapping(path = Array("/notificationRule"))
+  def createNotificationRule(notificationRule: NotificationRule)(implicit ec: ExecutionContext): CompletableFuture[NotificationRule] = {
```
Comment (jozefbakus) [inline src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/controllers/NotificationRuleController.scala]: Missing @RequestBody notificationRule: NotificationRule

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/controllers/NotificationRuleController.scala [hunk 6]
```diff
@@ -0,0 +1,63 @@
+/*
```
Comment (jozefbakus) [inline src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/controllers/NotificationRuleController.scala:1]: Missing
```
private class DagInstanceStatusesDeserializer extends JsonDeserializer[DagInstanceStatus] {
    override def deserialize(p: JsonParser, ctxt: DeserializationContext): DagInstanceStatus = {
      val node = p.getCodec.readTree[JsonNode](p)
      val value = node.get("name").textValue()
      DagInstanceStatuses.statuses.find(_.name == value).getOrElse(throw new Exception("Failed to find enum value"))
    }
  }
```

+ update: 
``` 
 private val module = new SimpleModule()
    .addDeserializer(classOf[SensorType], new SensorTypesDeserializer)
    .addDeserializer(classOf[JobStatus], new JobStatusesDeserializer)
    .addDeserializer(classOf[JobType], new JobTypesDeserializer)
    .addDeserializer(classOf[DagInstanceStatus], new DagInstanceStatusesDeserializer)
```
in Application.scala
Spring needs that for json deserialization

**Review by jozefbakus — DISMISSED** [review]

**Review by jozefbakus — APPROVED** [review]

**Pull request merged**